Enhanced Group Discovery / Normalized Names

NoMoreFood opened the pull request:
- Modified getusergroups() to use s4u tokens to discover nested groups and return them in NetBiosName\GroupName format.
- Modified get_passwd() to internally normalize names to NetBiosName\SamAccountName format and changed functions that use it to translate to UPN where necessary.
- Removed unnecessary support functions used by previous version of getusergroups().
- Various refactoring and function consolidation / simplification.
- Addressed several buffer over-read issues.

https://github.com/PowerShell/Win32-OpenSSH/issues/553

bagajjal commented on contrib/win32/win32compat/inc/grp.h:
```diff
@@ -3,14 +3,6 @@
-
 char ** getusergroups(const char *user, int *numgroups);
-void populate_user_groups(char **group_name, int *group_index, DWORD groupsread, DWORD totalgroups, LPBYTE buf, group_type groupType);
-void print_user_groups(const char *user, char **user_groups, int num_user_groups);
```
Have this function to print user groups. It will help in debugging.

NoMoreFood replied:
In latest commit, I ended up adding debug3() message for membership information.

NoMoreFood commented:
This is one of my original submissions so I know I have some style changes I need to make, but is there any other reason this hasn't been touched?

bagajjal commented on contrib/win32/win32compat/misc.c:
```diff
@@ -1115,216 +1098,192 @@ get_machine_domain_name(wchar_t *domain, int size)
-				wcscat_s(local_user_fmt_utf16, local_user_fmt_len, L"\\");
+	/* normalize the user input which should be specificied as username@fqdn */
+	char user_name_input[UNLEN + 1 + DNLEN + 1];
+	strcpy_s(user_name_input, sizeof(user_name_input), user);
```
By line 1148, wuser is in username@fqdn format
user is not modified so it will be either in user@domain or domain\user format?

bagajjal commented on contrib/win32/win32compat/misc.c:
```diff
@@ -1115,216 +1098,192 @@ get_machine_domain_name(wchar_t *domain, int size)
+	LUID LogonId;
+	QUOTA_LIMITS Quotas;
+	NTSTATUS SubStatus;
+	if ((LsaLogonUser(lsa_handle, &token_string, Network, lsa_authpackage, s4u_logon, s4u_size, NULL,
```
Please verify if you can reuse generate_s4u_user_token().

NoMoreFood replied:
Yes, I can use that, but I'll probably want to modify to accept a parameter that indicates we don't need a trusted LSA connection for simple group identification.  Question for you or @manojampalam.  It looks like generate_s4u_user_token() uses MsV1_0S4ULogon for local accounts which is not supported per the documentation: https://msdn.microsoft.com/en-us/library/windows/desktop/aa378764(v=vs.85).aspx  Do you guys have commitment from the security folks that this is appropriate and has long term support?

NoMoreFood replied:
Side note: generate_s4u_user_token() has a a bunch of small potential buffer overruns in the memcpy_s commands since we are copying beyond the potential bounds of string literals (e.g. copying 8 bytes out of "sshd" when the memory is points to is only guaranteed to have 5 bytes).  It's unlikely this would access memory we don't own, but I'll fix these if I have to touch these functions.

NoMoreFood replied:
Actually it looks like the support statement may just be ambiguous... I’m not if it’s telling us it’s not supported in Windows XP and earlier or later.

manojampalam replied:
Right. It is supposedly telling us that it is not supported Vista and below. I've emailed the owners to clarify it.

bagajjal commented on contrib/win32/win32compat/misc.c:
```diff
@@ -1115,216 +1098,192 @@ get_machine_domain_name(wchar_t *domain, int size)
-	if(!groupname) return NULL;
+	/* read group sids from logon token */
+	DWORD group_size = 0;
+	if (GetTokenInformation(logon_token, TokenGroups, NULL, 0, &group_size) == 0 && GetLastError() != ERROR_INSUFFICIENT_BUFFER ||
```
I am assuming TokenGroups will return both local groups and domain groups?

NoMoreFood replied:
Yes.

bagajjal commented on contrib/win32/win32compat/misc.c:
```diff
@@ -1115,216 +1098,192 @@ get_machine_domain_name(wchar_t *domain, int size)
+	DWORD group_size = 0;
+	if (GetTokenInformation(logon_token, TokenGroups, NULL, 0, &group_size) == 0 && GetLastError() != ERROR_INSUFFICIENT_BUFFER ||
+		GetTokenInformation(logon_token, TokenGroups, (PTOKEN_GROUPS)(group_buf =
+		(PTOKEN_GROUPS)malloc(group_size)), group_size, &group_size) == 0) {
```
As per [MSDN](https://msdn.microsoft.com/en-us/library/windows/desktop/aa446671(v=vs.85).aspx), if TokenInformation buffer is NULL then TokenInformationLength must be 0.

if malloc fails then we should pass 0 in place of group_size.

bagajjal commented on contrib/win32/win32compat/misc.c:
```diff
@@ -1115,216 +1098,192 @@ get_machine_domain_name(wchar_t *domain, int size)
+	/* open a new connection to the lsa policy provider for group lookup */
+	LSA_OBJECT_ATTRIBUTES ObjectAttributes;
+	ZeroMemory(&ObjectAttributes, sizeof(ObjectAttributes));
+	if (LsaOpenPolicy(NULL, &ObjectAttributes, POLICY_LOOKUP_NAMES, &lsa_policy) != 0) {
```
This requires running process as administrator.

Can you verify ssh using non-admin account (ssh non-admin-user@ip). I just want to make sure this works for both admin, non-admin users.

NoMoreFood replied:
Since we are using an untrusted LSA connection, it should work as non-admin.  The token generated cannot be used for user delegation -- only for obtaining information about the user.  I will certainly double-check using an untrusted user though.

NoMoreFood replied:
I verified this works fine as a non-admin.

bagajjal commented on contrib/win32/win32compat/misc.c:
```diff
@@ -1115,216 +1098,192 @@ get_machine_domain_name(wchar_t *domain, int size)
+		int current_group = (*ngroups)++;
+		PLSA_UNICODE_STRING domain = &(domain_list->Domains[name_list[group_index].DomainIndex].Name);
+		int name_size = snprintf(NULL, 0, "%wZ@%wZ", &(name_list[group_index].Name), domain) + 1;
+		sprintf_s((user_groups[current_group] = (char*)malloc(name_size)), name_size,
```
this code assumes malloc will not fail.
if malloc fails then it will be difficult to debug so please have a malloc NULL check.

NoMoreFood commented:
@bagajjal Please ignore this pull request for now.  I'm making some changes trying to more globally address some other login issues I'm seeing.

NoMoreFood commented:
@bagajjal @manojampalam  Alright, I reworked sys_auth_passwd(), get_passwd(), and getusergroups().  This takes care of the group nesting issues.  This should also address a few issues people reported.  SSH will now accept login formats as ‘DOMAIN\domainuser’, ‘domainuser’, ‘domainuser@domain.com’, ‘localuser’, and also internet identities bound to local accounts (e.g. Bryan.Berns@gmail.com).  This is all due to get_passwd() doing a better job normalizing the users and reconstructing the user string in sys_auth_passwd().  Previously you also could not do an AllowUser based on domain since the ‘@’ was being dropped in the matching algorithm.  My guess is that logins will also be a bit faster since there are fewer lookups on the server.
- The format for domain users in Deny/AllowUsers is ‘DOMAIN\user’.   Local users would just be ‘user’.  For example, DenyUsers DOMAIN\* would deny all users from domain 'DOMAIN'.
- The format for domain groups in Deny/AllowGroups is ‘DOMAIN\group’.  Local groups would just be ‘group’.

dwatley commented:
@NoMoreFood - Thanks for thorough contribution, excited to try this out. 👍

NoMoreFood commented on contrib/win32/win32compat/misc.c:
```diff
@@ -1115,216 +1102,165 @@ get_machine_domain_name(wchar_t *domain, int size)
+		DWORD rid = sid->SubAuthority[sid->SubAuthorityCount - 1]; 
+		SID_IDENTIFIER_AUTHORITY nt_authority = SECURITY_NT_AUTHORITY;
+		if (memcmp(&nt_authority, GetSidIdentifierAuthority(sid), sizeof(SID_IDENTIFIER_AUTHORITY)) == 0 && (
+			sub == SECURITY_NT_NON_UNIQUE || sub == SECURITY_BUILTIN_DOMAIN_RID) &&
```
Just wanted to note that the check (sub == SECURITY_BUILTIN_DOMAIN_RID) is to eliminate 'None' group from showing up (and causing the AppVeyor tests to fail that check for 0 group memberships).  The 'None' group is a real principal that's added to every local account (sub-authority: '-513').  When the accounts is domain joined, this resolves to 'Domain Users'.  You can actually assign permissions on the file system to 'None' and it's effectively the same as granting it to 'Authenticated Users'. It looks like the author of the 'whoami /groups' command actually tries to exclude this as well which has the side effect of making 'Domain Users' not show up on domain accounts (which it definitely should).  The ‘Users’ BUILTIN group is also always added to the token regardless of the user being in the ‘Users’ group.  Since these are all arguably real group principals that are used in group-based security checks, one could argue we should include them as well.  If we choose to do this though, we’d have to eliminate the “no groups” tests from AppVeyor since there’s really no such thing as a real user with no group principals -- the only way that would happen is S4U fails or name lookup would fails with would be hard to trigger in an E2E test.

manojampalam replied:
good to add this info as a comment in code  :)

NoMoreFood replied:
@manojampalam I can certainly expand the comment in the code, but I guess I meant this more as a question to you guys.  Since 'None' is a real group security principal that Windows will honor for other security checks within Windows, maybe we should eliminate this check altogether and include it in the group list?  I think that is the *right* thing to do and it would simplify the code, but I will have to eliminate the E2E tests that check for no groups since there won't be a realistic way to stage that test.  Let me know...

NoMoreFood commented:
One additional point that I didn't explain well.  Before this change, the output format from get_passwd() in the field pw.pw_name is all over the place.  This means that you can't reliably map AllowUsers/DenyUsers with a domain specified.  Examples:  

- Login Value: domainuser -> pw.pw_name: domainuser
- Login Value: domain\domainuser -> pw.pw_name: domainuser@domain
- Login Value: domainuser@domain.home.com -> pw.pw_name: domainuser@domain.home.com

Ideally, all of these should map to the same value.  Before this change, I'd have to specify all three in the AllowUsers filter for it to match consistently (domainuser, domainuser?domain, and domainuser?domain.home.com).  The unqualified name also presents a security concern since it could be a local user or it could be a domain user. 

In my proposal, they will all map to netbiosdomain\username for domain users and NetBiosDomain\SamAccountName.  I could change this to username@netbiosdomain, but then you have to resort to using the question mark as a matching character which is sort of ugly.

dwatley commented:
@NoMoreFood - So currently username@ad.corp.contoso.com maps to (?) username@contoso from what I see in the logs as the authenticated user and what works via pubkey auth principal match (noted in https://github.com/PowerShell/Win32-OpenSSH/issues/1055 and https://github.com/PowerShell/Win32-OpenSSH/issues/1061). Are you suggesting your change will modify that behavior or am I misunderstanding you and this explicitly only affects the AllowUsers/DenyUsers?

NoMoreFood commented:
@dwatley I'll verify the pubkey behavior with domains -- it probably affects that.  It'll affect anything that depends on using or decoding 'pw_name'.  AllowUsers/DenyUsers is one of those areas.  AllowUsers/DenyGroups is another area because it needs to properly reconstruct that the name so it can generate s4u tokens.   The ability for the user to login using a 'MicrosoftAccount' (i.e. internet identity) is also affected --- it works now.  

Looks like I have a conflict I need to resolve here so I'll take care of that tonight; also would love to hear back from @bagajjal to see if I can kill the SECURITY_BUILTIN_DOMAIN_RID check since I'm using it just to fake out and make the zero-group E2E tests pass even though the user actually does have active group security principals.

dwatley commented:
@NoMoreFood - Okay that's good to know. Your first case (unqualified domain user) presents an interesting scenario. What if there is an identical local user and domain user with the same name? What's the behavior?

manojampalam commented on auth-passwd.c:
```diff
@@ -282,12 +282,33 @@ sys_auth_passwd(struct ssh *ssh, const char *password)
+
+		/* if the domain looks fully qualified, put in user@domain format */
+		if (wcschr(udom_utf16, L'.') != NULL) {
+			wchar_t upn[1024 + 1];
```
Looks like this can be done irrespective of whether domain is fully-qualified/not. That way we can simplify and get rid of udom_utf16 altogether.

NoMoreFood replied:
I think this depends.  Do you consider passing username@netbios acceptable?  It may work empirically, but I don't think it's published as being a equivalent to UPN, is it?

manojampalam commented:
@NoMoreFood, first of all, thanks for thoroughly thinking through all the use case scenarios. I did go through all your changes related to user name resolution. I would like to get the design to be sorted and agreed on first.

For internet-connected accounts, I would like the username (pwd->pw_name) resolved to the local user name. Practically speaking, you'd see just he local user name everywhere from "net user" to "task manager" in any session associated with an internet-connected account. This will simplify the implementation very much, since the resolution would only need to take place as part of getpwd. You can remove duplicated "internet connected account logic" from within sys_auth_passwd and get_s4u_token .

As for the change in the name format (from user@domain to domain\user), the UPN was chosen since its the modern format. I do understand the benefit of using the legacy format ("?" issue). But changing this format is a pretty big breaking change now (since it changes the semantics of sshd_config user specs). I'm open to doing this breaking change, but want to be absolutely sure we do it once and right. I'll add in some folks from the community, lets take their opinion.

NoMoreFood commented:
@manojampalam I think local user name would make sense for internet identity --- it would simplify the S4U code I had to change as well.  The only downside is it will make AllowUser/DenyUser somewhat ambiguous since, in theory, I could have differently-named local accounts between computers for the same internet-connected account if a same-named account existed on the machine when the internet account was added.  However, I'm not sure this a practical concern.

I still think it would be good to switch to use the SamCompatibleName (NetBiosDomain\SamAccountName) for DomainUser.  I can actually have a SamCompatibleName like 'DOMAIN\Bryan' with a UPN for the same user of 'BryanBerns@mydomain.com'  (notice that DOMAIN != mydomain.com and Bryan != BryanBerns).  The released code actually does SamAccountName@NetBiosDomain when the user logs in with a SamCompatibleName which (as far as I know) follows no published standard.  My change would strictly follow the SamCompatibleName.   But yes, I understand this is a breaking change so I understand the apprehension... although I'm not sure how many folks are using the Allow/Deny rules yet.  Also, when it comes to groups, I think SamCompatibleName may be the only official option since groups don't have a UPN attribute --- so it would be nice to be consistent for everything.  Anyhow, let me know...

Note: I did change my 'pw_name' standardization proposal from UpnSuffix\UserName to SamCompatibleName (NetBiosDomain\SamAccountName) in one of the more recent commits.  This was actually done for performance since you really only need the UPN format for Kerberos S4U purposes and TranslateNameW does an AD query which can slow things down.

manojampalam commented:
@NoMoreFood yes and that would bring about my second part of the feedback. We are aware of the short comings of existing name resolution, we have work items to fix it, we haven't gotten to them yet - so appreciate your proactive efforts.

In my experience, I have seen issues dealing with NetBios names, specifically in a forest of inter-trusted domains - say domA.parent.com and domB.parent.com. On a system joined to domA, while interating with domB.parent.com with just domB will work most of the time - it would occasionally fail and the immediate enduser work around would be fully qualify domB (as domB.parent.com). I believe @dwatley hit this recently.

I can imagine similar situation dealing with domain accounts over OpenSSH. I think it would be wise to always resolve domain user accounts with FQDNs (whether its domB.parent.com/user or user@domB.parent.com). Since, if we dont do this, we may still be at the mercy of underlying resolution that may fail (say while generating s4u token). This approach, however, has a cons - admins now have to deal with FQDNs with user principals in sshd_config. Ex
`MatchUser child.parent.com\domainuser`
and this may become inconvenient as the tree grows. But, I beieve, this can easily be overcome using wildcard matching.
`MatchUser child.*\domainuser`

NoMoreFood committed:
Enhanced Group Discovery / Normalized Names

- Modified getusergroups() to use s4u tokens to discover nested groups and return them in NetBiosName\GroupName format.
- Modified get_passwd() to internally normalize names to NetBiosName\SamAccountName format and changed functions that use it to translate to UPN where necessary.
- Removed unnecessary support functions used by previous version of getusergroups().
- Various refactoring and function consolidation / simplification.
- Addressed several buffer over-read issues.

NoMoreFood commented:
@manojampalam Alright, the my latest commit (which I squashed at the other commits) uses the resolved local name for internet-linked identities.  This simplified some of the parsing downstream.  I'm still using NetBiosDomain\SamAccountName for the pw_nam field and which is attempted to be converted to use UserPrincipalName.  I thought of another reason we might not want to do UserPrincipalName: UserPrincipalName is not a mandatory attribute for AD (I actually hit this in my testing).  It's usually populated, but not always.  So the best we could do that'll always be available is NetBiosDomain\SamAccountName or the less familiar NameDnsDomain (as defined here https://msdn.microsoft.com/en-us/library/windows/desktop/ms724268(v=vs.85).aspx).  Using NameDnsDomain, however, would still require us to conversion to UserPrincipalName in the S4U code since the UserPrincipalName suffix may very well be different than NameDnsDomain for a domain with children.  We'd also have to add additional conversion from NameDnsDomain to NetBiosDomain where NetBiosDomain is necessary (e.g., loading profiles).

manojampalam commented:
@NoMoreFood Perhaps, I didn't get my proposal clear. I'm proposing we resolve to this
`domain.home.com\domainuser  (FQDN\username)`
instead of
`domain\domainuser`
Let me know if I'm missing something.

dwatley commented:
@manojampalam - My issue wasn't in a multi-domain environment, but rather an environment that has no NetBIOS resolution available. This caused the initial timeout for NetUserGetInfo. Using the FQDN (username@ad.corp.contoso.com) was the workaround there. This differs from the format that is required on the principal (username@contoso) for pubkey auth. Unless I've misunderstood, this PR will change this behavior and would be considered a breaking change, which would mean the community may have to re-engineer systems that made certain assumptions. So understanding the potential impact would be good. "once and right" I agree, if the pros outweigh the cons?

manojampalam commented:
@dwatley thanks for the clarification. 

This PR is going to be a breaking change in all scenarios that deal with user name identities including but not limited to:
- Match blocks (in sshd_config)
- Certificate authentication

I'm proposing we resolve all domain accounts with FQDNs (domain.home.com\domainuser) while the above changes from @NoMoreFood resolve them to domain\domainuser.

NoMoreFood commented:
@manojampalam I can make the change, but I just wanted to be sure you wanted you absolutely want NameDnsDomain format.  I don't really think it gives us anything and requires an additional TranslateName() upfront to normalize the name to NameDnsDomain and then additional TranslateName() conversions later to convert NameDnsDomain to UserPrincipalName (since DnsDomain is not necessarily equal to UserPrincipalName suffix) for logon/s4u and NetBiosName\SamAccountName for profile loading, etc.  I think it's much more apt to produce problems than solve them.  I believe @dwatley issues were related to the fact that NetBiosName\SamAccountName is being converted to SamAccountName@NetBiosName in the released code which likely is making the system attempt to contact NetBiosName as an address. Translating NetBiosName\SamAccountName via TranslateName() to other formats does not utilize NetBios name resolution so I'm almost certain these changes would work for him.

Using NetBiosName\SamAccountName for pw_nam (with conversion to UserPrincipalName in s4u/logon), I've successfully tested a variety of input formats using using cross-domain accounts and cross-forest accounts. This was actually all done with NetBios name resolution turned off and using NetBios domain names that don't happen to be resolvable via DNS since they are actually different from the domain (e.g. NetBios: CHILDNB; Domain: child.domain.home.com).  I did this for both host-based login and password-based login.  My day job is being an Enterprise Administrator for large enterprise so my home test lab is well equipped ;-) 

So given NameDnsDomain is a somewhat obscure format and actually requires more conversions, I just wanted to be sure this is actually the route you want to go.

manojampalam commented:
@NoMoreFood thanks for the follow up. Let me do a bit more homework on my side and get back.

NoMoreFood commented:
@manojampalam Sounds good.

NoMoreFood committed:
Enhanced Group Discovery / Normalized Names - Revisions

- Moved return structure declaration to ensure proper return value on failure.

manojampalam commented:
@NoMoreFood check out this scenario:
2 different user identies with same username and short domain names but belonging to different domains:
administrator\example.com
administrator\example.example.com
I believe these 2 are valid UPNs (though may not be practical).

In such cases, full FQDN would typically be specified for unique identification. If we drop the FQDN in our resolution and include only short name (or NetBios name or hostname), we are ending up dropping information that would uniquely identify the client identity.

NoMoreFood commented:
Two domains in the same forest cannot have the same NetBios name -- Active Directory won't let you do that.  I believe it is technically possible to have that sort of scenario between child domains in two forests that have trusts between them, but I have never encountered this and other AD-aware apps could have issues.  The most popular SID to Name resolution function (LookupAccountSid) essentially converts SIDs to the NetBios\SamAccountName format -- as far as I know there is no equivalent function that elegantly goes from SID to NameDnsDomain format (would be nice if their was).  That's probably why NetBios\SamAccountName is still used for name display in so many places within Windows.

You mentioned UPN in the last post so I just wanted to again clarify that UserPrincipalName suffix is not the same DnsDomain from the NameDnsDomain format.  For example, in my company forest I would have approximately 200 users matching NameDnsDomain for "example.com\\\*" whereas I have more than 20,000 users that have a UserPrincialName matching "*@example.com".  This is because users in child domains (e.g., child.example.com\\\*) can optionally use the parent domain suffix for the UserPrincialName suffix (@example.com).  In fact, you can add custom UserPrincialName suffixes in your forest that have nothing to do with your forest fully qualified names.  And accounts can optionally not have UserPrincialName defined at all.  Apologize if all this is understood, I just want to be sure.

So if we use NameDnsDomain format, we just to be sure to educate users that it's alternate format for  NetBios\\SamAccountName and not a string permutation of UserPrincipalName. Also, I'm seeing that the command prompt shell looks like it puts pw_nam as the default prompt and its limited to 32 characters -- not sure what's doing that but if we use the NameDnsDomain, we may also want to considered elegantly truncating that in some way.

NoMoreFood commented:
Any further thoughts on this?  I'm happy to accommodate whatever you decide.  I think the most important thing is to normalize the names, whatever the format might be.

manojampalam commented:
Not yet. I'm following up with some folks internally to ensure we do the right thing. Thanks for your patience.

dwatley commented:
/cc @djmdjm

manojampalam commented:
Alright. Lets proceed with NameSamCompatible format. I'll finish up reviewing rest of the code.

manojampalam commented on auth-passwd.c:
```diff
@@ -272,22 +274,46 @@ int
-	    (pwd_utf16 = utf8_to_utf16(password)) == NULL) {
-		fatal("out of memory");
+	wchar_t * user_utf16 = utf8_to_utf16(authctxt->pw->pw_name);
+	wchar_t * pwd_utf16 = utf8_to_utf16(password);
```
Move all declarations up. This is in common source, lets keep it consistent with rest of the code.

manojampalam commented on contrib/win32/win32compat/pwd.c:
```diff
@@ -99,213 +103,145 @@ reset_pw()
+	/* attempt to lookup the account; this will verify the account is valid and
+	* is will return its sid and the realm that owns it */
+	else if(LookupAccountNameW(NULL, user_utf16, binary_sid, &sid_size,
+		domain_name, &domain_name_size, &account_type) == 0) {
```
pass NULL for domain_name. We don't intend on using the domain output of this function.

NoMoreFood replied:
The LookupAccountName() function documentation is deceptive about the "optional" nature of this parameter.  It's only optional so far as you don't have to specify it when obtaining the size.  If you pass in NULL with the size set to 0, the function returns 0 (failure) with an GetLastError() = ERROR_INSUFFICIENT_BUFFER, and the only output parameter that's valid is the domain size; no other data is populated.

manojampalam replied:
ok. good to know.

manojampalam commented on contrib/win32/win32compat/pwd.c:
```diff
@@ -99,213 +103,145 @@ reset_pw()
+	/* verify passed account is actually a user account */
+	if (account_type != SidTypeUser) {
+		errno = ENOENT;
+		debug3("%s: Invalid account type.", __FUNCTION__);
```
log account type

manojampalam commented on contrib/win32/win32compat/pwd.c:
```diff
@@ -99,213 +103,145 @@ reset_pw()
-	    ExpandEnvironmentStringsW(profile_home, NULL, 0) > PATH_MAX || 
-	    ExpandEnvironmentStringsW(profile_home, profile_home_exp, PATH_MAX) == 0)
+	if (swprintf_s(reg_path, PATH_MAX, L"SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion\\ProfileList\\%ls", sid_string) == -1 ||
+		RegOpenKeyExW(HKEY_LOCAL_MACHINE, reg_path, 0, STANDARD_RIGHTS_READ | KEY_QUERY_VALUE | KEY_WOW64_64KEY, &reg_key) != 0 ||
```
Replace these preceeding tabs with 4 spaces (conditional sections` indentation rules from OpenBSD coding guidelines)

manojampalam commented on contrib/win32/win32compat/win32_usertoken_utils.c:
```diff
@@ -87,142 +92,131 @@ EnablePrivilege(const char *privName, int enabled)
 	if (domain_user) {
+
+		/* lookup the upn for the user */
+		WCHAR domain_upn[1024 + 1];
```
replace 1024 with MAX* macros

manojampalam commented on contrib/win32/win32compat/win32_usertoken_utils.c:
```diff
@@ -288,7 +282,7 @@ process_custom_lsa_auth(const char* user, const char* pwd, const char* lsa_pkg)
 		goto done;
```
Need to fix domain related logic in process_custom_lsa_auth(). It still assumes the older CPN like format.

NoMoreFood replied:
I hadn't looked at this code much before.  It looks like it allows an administrator-provided authentication package by name (via registry key) and then passes in the login information as "username;password" or "username;password;domain".  Based on quick read of the APIs, it looks like this logon info format should be package-dependent.   Was the intent her to target a specific authentication package?  If so, which one since I'd like to test?  If not, is this logon info format documented somewhere as being a standard that custom authentication packages should use?

manojampalam replied:
You reverse engineered it right. @bagajjal can you document how this authentication extension can be used.

NoMoreFood replied:
@bagajjal I suppose I really don't need a verbose write-up and/or test it myself, but it would be useful to know what the ideal username/domain format is.  For example, is it SamAccountName;Password:NetBiosName, SamAccountName;Password;DnsDomainName, UpnAccountName;Password;UpnAccountSuffix, or something else entirely?

bagajjal replied:
its "username:password:domain".
Here are the steps to test this,
1) download, unzip the [sample lsa provider] (https://github.com/PowerShell/openssh-portable/files/1944211/lsa-provider.zip)
2) from elevated powershell window run ".\install-lsaprov.ps1"
3) Add key "HKLM\SOFTWARE\OpenSSH\LSAAuthenticationPackage" with value "custompwdauth-lsa"
![image](https://user-images.githubusercontent.com/23668037/39208752-b7bff26a-47b8-11e8-8662-a9d26528857d.png)
4) Restart computer.
5) login using ssh - ssh user@ip. Enter dummy password, it should let you login using custom lsa provider configured in step 2.

NoMoreFood replied:
So is this meant to be a theoretical custom provider that somebody could custom-code just for SSH?  Unless I'm missing something, right now the code you sent basically only uses the first token in the "username;password;domain" format. The password and domain parameters aren't used. Moreover, it expects that the username to be in the UserPrincipalName format (aka SecNameFlat for GetAuthDataForUser) which would make the domain field unnecessary (if it were actually being used).  Can you provide some additional context as to what we are actually trying to accomplish with this provider?

bagajjal replied:
@NoMoreFood - This is just a sample code used for testing. We have an internal requirement to support custom lsa authentication provider in "username:password:domain" format.

NoMoreFood replied:
@bagajjal Ok, that makes a bit more sense.  The current code actually uses semicolons, not colons and the LSA sample tokenizes on semicolon.  So is it semicolon or colon?  Also, by domain do you mean NetBiosName (DOMAIN) or a fully qualified name (domain.com)?

bagajjal replied:
My bad its semicolon, "username;password;domain". Domain is NetBiosName (DOMAIN).

NoMoreFood replied:
@bagajjal My changes are in for this.

dwatley commented:
@manojampalam - So `NameSamCompatible` == NetBios\SamAccountName ? Does the slash for the principal need to be escaped in a user cert or no? contoso\\\dwatley ?

manojampalam commented:
@dwatley why would you escape the slash ?

This should now work:
ssh-keygen.exe -s id_ca -I cert_identity -V -1w:+54w5d -n contoso\dwatley id_user.pub

dwatley commented:
Just noticed in the logs somewhere before seeing double slashes for an identity. I'll give that a shot whenever this is merged and released, but acknowledging the format change.

NoMoreFood committed:
Enhanced Group Discovery / Normalized Names - Revisions 2

- Minor changes to resolve several comments in first round of review.

NoMoreFood committed:
Enhanced Group Discovery / Normalized Names - Revisions 3

- Updated process_custom_lsa_auth() to utilize updated user format and use variable case consistent with other functions.
- Updated getusergroups() to use built-in macros and corrected a comment.

NoMoreFood committed:
Merge branch 'latestw_all' into group_enhancement

manojampalam commented on contrib/win32/win32compat/win32_usertoken_utils.c:
```diff
@@ -316,12 +293,12 @@ process_custom_lsa_auth(const char* user, const char* pwd, const char* lsa_pkg)
+	if (user_utf16)
+		free(user_utf16);
+	if (pwd_utf16)
+		free(pwd_utf16);
```
use SecureZeroMemory for passwords

NoMoreFood replied:
Done.

NoMoreFood committed:
Enhanced Group Discovery / Normalized Names - Revisions 4

- Updated process_custom_lsa_auth() to securely zero password data.

NoMoreFood commented:
Please let me know if you think there's anything outstanding for comments.
